File: Week5 (Tree-2 [BST, MST], Heap, Red-Black Tree)/Red-Black Tree/khj_RBT.py
'''
1. 모든 Node의 색상은 Red or Black
2. 루트 Node는 무조건 Black
3. 모든 Leaf 노드(Nil 노드라고 한다)는 무조건 Black
4. Red 노드의 자식 노드 양쪽은 언제나 Black (== 빨간색 노드가 연속으로 나올 수 없다)
5. 어떤 노드에서 Leaf 노드까지의 경로들은 모두 같은 수의 Black 노드를 포함하고 있다.
'''

#이진 트리 구현한 노드에서 color 값만 추가됨
class RBNode:
    def __init__(self, key) :
        self.key = key
        self.color = "RED"

        self.left = None
        self.right = None
        self.parent = None

    def __str__(self):
        return str(self.key)
    

# EXT 라는 멤버를 두어 RBNode 인스턴스 하나를 참조하게 함. 외부 노드이므로 black
class RedBlackTree:
    def __init__(self):
        self.__root = None  #루트 노드
        self.__EXT = RBNode(None)  # 외부 노드로 사용할 특별한 노드 생성
        self.__EXT.color = "BLACK"  # 외부 노드는 항상 black

    def get_root(self):
        return self.__root
    
    def preorder_traverse(self, cur, func, *args, **kwargs):
        if cur == self.__EXT:  #외부 노드에 도달하면 반환
            return   
        
        func(cur, *args, **kwargs)     # 현재 노드에 대한 함수 실행
        self.preorder_traverse(cur.left, func, *args, **kwargs)
        self.preorder_traverse(cur.right, func, *args, **kwargs)


    #함수 오름차순 방문
    def inorder_traverse(self, cur, func, *args, **kwargs):
        if cur == self.__EXT:
            return
        self.inorder_traverse(cur.left, func, *args, **kwargs)
        func(cur, *args, **kwargs)
        self.inorder_traverse(cur.right, func, *args, **kwargs)


    # 트리 높이 계산
    def tree_height(self, node):
        if node == self.__EXT:
            return -1  # 외부 노드는 높이에 포함되지 않음
        left_height = self.tree_height(node.left)
        right_height = self.tree_height(node.right)
        return max(left_height, right_height) + 1


    # 트리 노드 개수 계산
    def node_count(self, node):
        if node == self.__EXT:
            return 0
        return 1 + self.node_count(node.left) + self.node_count(node.right)



    # 특정 노드의 black height 계산
    def __black_height(self, node):
        if node == self.__EXT:
            return 0
        left_black_height = self.__black_height(node.left)
        right_black_height = self.__black_height(node.right)
        if node.color == "BLACK":
            return max(left_black_height, right_black_height) + 1
        else:
            return max(left_black_height, right_black_height)




    #왼쪽으로 회전
    '''
    회전 기준 노드 x의 오른쪽 자식 y가 새로운 부모 노드가 됩니다.
    x는 y의 왼쪽 자식이 됩니다.
    y의 왼쪽 자식 T2는 x의 오른쪽 자식이 됩니다.
    '''
    def __left_rotate(self, n):
            r = n.right #n의 오른쪽 자식
            l = r.left #r의 왼쪽 자식
  
            l.parent = n   #l의 부모를 n으로 설정
            n.right = l    #l은 n의 오른쪽 자식


            if n == self.__root:
                self.__root = r          #n이 루트 노드인 경우 r을 새로운 루트로 설정
            elif n.parent.left == n:
                n.parent.left = r
            else:
                n.parent.right = r
            
            r.parent = n.parent     # r의 부모를 n의 부모로 설정
            r.left = n       # r의 왼쪽 자식을 n으로 설정
            n.parent = r        




    # 오른쪽으로 회전
    '''
    회전 기준 노드 x의 왼쪽 자식 y가 새로운 부모 노드가 됩니다.
    x는 y의 오른쪽 자식이 됩니다.
    y의 오른쪽 자식 T2는 x의 왼쪽 자식이 됩니다.
    '''
    def __right_rotate(self, n):
            l = n.left
            r = l.right

            r.parent = n
            n.left = r

            if n == self.__root:
                self.__root = r
            elif n.parent.left == n:
                n.parent.left = l
            else:
                n.parent.right = l

            l.parent = n.parent
            l.right = n
            n.parent = l



    # insert 연산 : while 문 이후 insert_fix 연산 수행하여 노드의 재배치 수행
    def insert(self, key):
            new_node = RBNode(key)
            new_node.left=self.__EXT
            new_node.right=self.__EXT

            cur = self.__root

            if not cur:
                self.__root = new_node          # 트리가 비어있으면 새 노드를 루트로 설정
                self.__root.color = "BLACK"
                return

            while True:
                parent = cur
                if key < cur.key:
                    parent.left = new_node
                    new_node.parent = parent
                    break
                else:
                    cur = cur.right
                    if cur == self.__EXT:
                        parent.right = new_node
                        new_node.parent = parent
                        break

            # 삽입 후 균형을 맞추기 위해 수정
            self.__insert_fix(new_node)



    def __insert_fix(self, n):
            pn = gn = un = None
            pn = n.parent #n의 부모

            while pn != None and pn.color == "RED":  #pn의 color가 red면 반드시 조부모 존재
                gn = pn.parent  #gn은 n의 조부모

                if gn.left == pn:
                    un = gn.right  #pn의 형제

                    if un.color == "RED":   
                        gn.color = "RED"
                        pn.color = un.color = "BLACK"

                        n = gn
                        pn = n.parent

                    else:
                        if pn.right == n:
                            self.__left_rotate(pn)  # 부모를 기준으로 왼쪽 회전
                            n, pn = pn, n

                        pn.color, gn.color = gn.color, pn.color

                    self.__right_rotate(gn)
                
                
                else:
                    un = gn.left
                    if un.color == "RED":
                        gn.color = "RED"
                        pn.color = un.color = "BLACK"

                        n = gn
                        pn = n.parent

                    else:     # 부모가 조부모의 오른쪽 자식인 경우
                        # n이 부모의 왼쪽 자식인 경우
                        if pn.left == n:
                            self.__right_rotate(pn)

                        # n이 부모의 오른쪽 자식인 경우    
                        n, pn = pn, n
                        pn.color, gn.color = gn.color, pn.color
                        self.__left_rotate(gn)

            self.__root.color = "BLACK" #루트 노드는 항상 black


    def search(self, key):
        cur = self.__root
        while cur != self.__EXT and cur.key != key:
            if key < cur.key:
                cur = cur.left
            else:
                cur = cur.right
        return cur



    def delete(self, key):
        node = self.search(key)
        if node == self.__EXT:
            return  # 노드가 없으면 삭제할 것이 없음

        y = node
        y_original_color = y.color
        if node.left == self.__EXT:
            x = node.right
            self.__transplant(node, node.right)
        elif node.right == self.__EXT:
            x = node.left
            self.__transplant(node, node.left)
        else:
            y = self.__minimum(node.right)
            y_original_color = y.color
            x = y.right
            if y.parent == node:
                x.parent = y
            else:
                self.__transplant(y, y.right)
                y.right = node.right
                y.right.parent = y
            self.__transplant(node, y)
            y.left = node.left
            y.left.parent = y
            y.color = node.color
        if y_original_color == "BLACK":
            self.__delete_fix(x)



    def __delete_fix(self, x):
        while x != self.__root and x.color == "BLACK":
            if x == x.parent.left:
                w = x.parent.right
                if w.color == "RED":
                    w.color = "BLACK"
                    x.parent.color = "RED"
                    self.__left_rotate(x.parent)
                    w = x.parent.right
                if w.left.color == "BLACK" and w.right.color == "BLACK":
                    w.color = "RED"
                    x = x.parent
                else:
                    if w.right.color == "BLACK":
                        w.left.color = "BLACK"
                        w.color = "RED"
                        self.__right_rotate(w)
                        w = x.parent.right
                    w.color = x.parent.color
                    x.parent.color = "BLACK"
                    w.right.color = "BLACK"
                    self.__left_rotate(x.parent)
                    x = self.__root
            else:
                w = x.parent.left
                if w.color == "RED":
                    w.color = "BLACK"
                    x.parent.color = "RED"
                    self.__right_rotate(x.parent)
                    w = x.parent.left
                if w.right.color == "BLACK" and w.left.color == "BLACK":
                    w.color = "RED"
                    x = x.parent
                else:
                    if w.left.color == "BLACK":
                        w.right.color = "BLACK"
                        w.color = "RED"
                        self.__left_rotate(w)
                        w = x.parent.left
                    w.color = x.parent.color
                    x.parent.color = "BLACK"
                    w.left.color = "BLACK"
                    self.__right_rotate(x.parent)
                    x = self.__root
        x.color = "BLACK"



    def __transplant(self, u, v):
        if u.parent is None:
            self.__root = v
        elif u == u.parent.left:
            u.parent.left = v
        else:
            u.parent.right = v
        v.parent = u.parent


    def __minimum(self, node):
        while node.left != self.__EXT:
            node = node.left
        return node

    def minimum(self):
        return self.__minimum(self.__root)

    def __maximum(self, node):
        while node.right != self.__EXT:
            node = node.right
        return node

    def maximum(self):
        return self.__maximum(self.__root)



    def print_node(self, rbn):
        if rbn:
            print("node : {}, ".format(rbn.key), end=" ")
            if rbn.color == "RED":
                print("color : RED, ", end=" ")
            else:
                print("color : BLACK, ", end=" ")
            if rbn.left:
                print("left : {}, ".format(rbn.left.key), end=" ")
            if rbn.right:
                print("right : {}, ".format(rbn.right.key), end=" ")
            if rbn.parent:
                print("parent : {}".format(rbn.parent.key), end=" ")
            print()




if __name__ == "__main__":
    print('*'*100)
    rbt = RedBlackTree()

    for i in range(10):
        rbt.insert(i)

    rbt.preorder_traverse(rbt.get_root(), rbt.print_node)
    print('*'*100)

    # 삭제 테스트
    rbt.delete(3)
    rbt.delete(7)

    rbt.preorder_traverse(rbt.get_root(), rbt.print_node)
    print('*'*100)

File: Week5 (Tree-2 [BST, MST], Heap, Red-Black Tree)/Red-Black Tree/khj_RBT.cpp
#include <iostream>
#include <algorithm>
#include <vector>
#include <functional>

using namespace std;

enum Color { RED, BLACK };

// RBNode 클래스
class RBNode {
public:
    int key;
    Color color;
    RBNode* left;
    RBNode* right;
    RBNode* parent;

    RBNode(int key) : key(key), color(RED), left(nullptr), right(nullptr), parent(nullptr) {}

    friend ostream& operator<<(ostream& os, const RBNode& node) {
        os << node.key;
        return os;
    }
};

class RedBlackTree {
private:
    RBNode* root;
    RBNode* EXT; // 외부 노드

    void left_rotate(RBNode* x) {
        RBNode* y = x->right; // x의 오른쪽 자식
        x->right = y->left; // y의 왼쪽 자식 T2를 x의 오른쪽 자식으로 설정

        if (y->left != EXT) {
            y->left->parent = x; // T2의 부모를 x로 설정
        }

        y->parent = x->parent; // y의 부모를 x의 부모로 설정

        if (x->parent == nullptr) {
            root = y; // x가 루트 노드인 경우 y를 새로운 루트로 설정
        } else if (x == x->parent->left) {
            x->parent->left = y; // x가 왼쪽 자식인 경우 y를 x의 부모의 왼쪽 자식으로 설정
        } else {
            x->parent->right = y; // x가 오른쪽 자식인 경우 y를 x의 부모의 오른쪽 자식으로 설정
        }

        y->left = x; // x를 y의 왼쪽 자식으로 설정
        x->parent = y; // x의 부모를 y로 설정
    }

    void right_rotate(RBNode* x) {
        RBNode* y = x->left; // x의 왼쪽 자식
        x->left = y->right; // y의 오른쪽 자식 T2를 x의 왼쪽 자식으로 설정

        if (y->right != EXT) {
            y->right->parent = x; // T2의 부모를 x로 설정
        }

        y->parent = x->parent; // y의 부모를 x의 부모로 설정

        if (x->parent == nullptr) {
            root = y; // x가 루트 노드인 경우 y를 새로운 루트로 설정
        } else if (x == x->parent->right) {
            x->parent->right = y; // x가 오른쪽 자식인 경우 y를 x의 부모의 오른쪽 자식으로 설정
        } else {
            x->parent->left = y; // x가 왼쪽 자식인 경우 y를 x의 부모의 왼쪽 자식으로 설정
        }

        y->right = x; // x를 y의 오른쪽 자식으로 설정
        x->parent = y; // x의 부모를 y로 설정
    }

    void insert_fix(RBNode* n) {
        RBNode* pn = nullptr;
        RBNode* gn = nullptr;
        RBNode* un = nullptr;

        while (n->parent != nullptr && n->parent->color == RED) {
            pn = n->parent; // n의 부모
            gn = pn->parent; // n의 조부모

            if (pn == gn->left) {
                un = gn->right; // 삼촌 노드

                if (un != nullptr && un->color == RED) {
                    // Case 1: 삼촌이 RED인 경우
                    gn->color = RED;
                    pn->color = BLACK;
                    un->color = BLACK;
                    n = gn; // 조부모를 기준으로 다시 균형 맞추기
                } else {
                    if (n == pn->right) {
                        // Case 2: n이 부모의 오른쪽 자식인 경우
                        left_rotate(pn);
                        n = pn;
                        pn = n->parent;
                    }

                    // Case 3: n이 부모의 왼쪽 자식인 경우
                    pn->color = BLACK;
                    gn->color = RED;
                    right_rotate(gn);
                }
            } else {
                un = gn->left; // 삼촌 노드

                if (un != nullptr && un->color == RED) {
                    // Case 1: 삼촌이 RED인 경우
                    gn->color = RED;
                    pn->color = BLACK;
                    un->color = BLACK;
                    n = gn; // 조부모를 기준으로 다시 균형 맞추기
                } else {
                    if (n == pn->left) {
                        // Case 2: n이 부모의 왼쪽 자식인 경우
                        right_rotate(pn);
                        n = pn;
                        pn = n->parent;
                    }

                    // Case 3: n이 부모의 오른쪽 자식인 경우
                    pn->color = BLACK;
                    gn->color = RED;
                    left_rotate(gn);
                }
            }
        }

        root->color = BLACK; // 루트 노드는 항상 BLACK
    }

    void transplant(RBNode* u, RBNode* v) {
        if (u->parent == nullptr) {
            root = v;
        } else if (u == u->parent->left) {
            u->parent->left = v;
        } else {
            u->parent->right = v;
        }
        if (v != nullptr) {
            v->parent = u->parent;
        }
    }

    RBNode* minimum(RBNode* node) {
        while (node->left != EXT) {
            node = node->left;
        }
        return node;
    }

    void delete_fix(RBNode* x) {
        while (x != root && x->color == BLACK) {
            if (x == x->parent->left) {
                RBNode* w = x->parent->right;
                if (w->color == RED) {
                    w->color = BLACK;
                    x->parent->color = RED;
                    left_rotate(x->parent);
                    w = x->parent->right;
                }
                if (w->left->color == BLACK && w->right->color == BLACK) {
                    w->color = RED;
                    x = x->parent;
                } else {
                    if (w->right->color == BLACK) {
                        w->left->color = BLACK;
                        w->color = RED;
                        right_rotate(w);
                        w = x->parent->right;
                    }
                    w->color = x->parent->color;
                    x->parent->color = BLACK;
                    w->right->color = BLACK;
                    left_rotate(x->parent);
                    x = root;
                }
            } else {
                RBNode* w = x->parent->left;
                if (w->color == RED) {
                    w->color = BLACK;
                    x->parent->color = RED;
                    right_rotate(x->parent);
                    w = x->parent->left;
                }
                if (w->right->color == BLACK && w->left->color == BLACK) {
                    w->color = RED;
                    x = x->parent;
                } else {
                    if (w->left->color == BLACK) {
                        w->right->color = BLACK;
                        w->color = RED;
                        left_rotate(w);
                        w = x->parent->left;
                    }
                    w->color = x->parent->color;
                    x->parent->color = BLACK;
                    w->left->color = BLACK;
                    right_rotate(x->parent);
                    x = root;
                }
            }
        }
        x->color = BLACK;
    }

public:
    RedBlackTree() {
        EXT = new RBNode(-1);
        EXT->color = BLACK;
        root = EXT;
    }

    RBNode* get_root() {
        return root;
    }

    void preorder_traverse(RBNode* cur, function<void(RBNode*)> func) {
        if (cur == EXT) return;
        func(cur);
        preorder_traverse(cur->left, func);
        preorder_traverse(cur->right, func);
    }

    void inorder_traverse(RBNode* cur, function<void(RBNode*)> func) {
        if (cur == EXT) return;
        inorder_traverse(cur->left, func);
        func(cur);
        inorder_traverse(cur->right, func);
    }


    int tree_height(RBNode* node) {
        if (node == EXT) return -1;
        int left_height = tree_height(node->left);
        int right_height = tree_height(node->right);
        return max(left_height, right_height) + 1;
    }

    int node_count(RBNode* node) {
        if (node == EXT) return 0;
        return 1 + node_count(node->left) + node_count(node->right);
    }

    int black_height(RBNode* node) {
        if (node == EXT) return 0;
        int left_black_height = black_height(node->left);
        int right_black_height = black_height(node->right);
        if (node->color == BLACK) {
            return max(left_black_height, right_black_height) + 1;
        } else {
            return max(left_black_height, right_black_height);
        }
    }

    void insert(int key) {
        RBNode* new_node = new RBNode(key);
        new_node->left = EXT;
        new_node->right = EXT;

        RBNode* cur = root;

        if (root == EXT) {
            root = new_node;
            root->color = BLACK;
            root->parent = nullptr;
            return;
        }

        RBNode* parent = nullptr;

        while (cur != EXT) {
            parent = cur;
            if (key < cur->key) {
                cur = cur->left;
            } else {
                cur = cur->right;
            }
        }

        new_node->parent = parent;
        if (key < parent->key) {
            parent->left = new_node;
        } else {
            parent->right = new_node;
        }

        insert_fix(new_node);
    }

    RBNode* search(int key) {
        RBNode* cur = root;
        while (cur != EXT && cur->key != key) {
            if (key < cur->key) {
                cur = cur->left;
            } else {
                cur = cur->right;
            }
        }
        return cur;
    }

    void delete_node(int key) {
        RBNode* node = search(key);
        if (node == EXT) return;

        RBNode* y = node;
        RBNode* x;
        Color y_original_color = y->color;

        if (node->left == EXT) {
            x = node->right;
            transplant(node, node->right);
        } else if (node->right == EXT) {
            x = node->left;
            transplant(node, node->left);
        } else {
            y = minimum(node->right);
            y_original_color = y->color;
            x = y->right;
            if (y->parent == node) {
                x->parent = y;
            } else {
                transplant(y, y->right);
                y->right = node->right;
                y->right->parent = y;
            }
            transplant(node, y);
            y->left = node->left;
            y->left->parent = y;
            y->color = node->color;
        }

        delete node;

        if (y_original_color == BLACK) {
            delete_fix(x);
        }
    }

    void print_node(RBNode* node) {
        if (node) {
            cout << "node: " << node->key << ", ";
            if (node->color == RED) {
                cout << "color: RED, ";
            } else {
                cout << "color: BLACK, ";
            }
            if (node->left) {
                cout << "left: " << node->left->key << ", ";
            }
            if (node->right) {
                cout << "right: " << node->right->key << ", ";
            }
            if (node->parent) {
                cout << "parent: " << node->parent->key;
            }
            cout << endl;
        }
    }
};


int main() {
    RedBlackTree rbt;

    for (int i = 0; i < 10; ++i) {
        rbt.insert(i);
    }

    rbt.preorder_traverse(rbt.get_root(), [&](RBNode* node) { rbt.print_node(node); });
    cout << string(100, '*') << endl;

    // 삭제 테스트
    rbt.delete_node(3);
    rbt.delete_node(7);

    rbt.preorder_traverse(rbt.get_root(), [&](RBNode* node) { rbt.print_node(node); });
    cout << string(100, '*') << endl;

    return 0;
}

